### PR TITLE
[SPARK-17874][core] Add SSL port configuration.

### PR DESCRIPTION
**Pull request by vanzin**
Make the SSL port configuration explicit, instead of deriving it
from the non-SSL port, but retain the existing functionality in
case anyone depends on it.

The change starts the HTTPS and HTTP connectors separately, so
that it's possible to use independent ports for each. For that to
work, the initialization of the server needs to be shuffled around
a bit. The change also makes it so the initialization of both
connectors is similar, and end up using the same Scheduler - previously
only the HTTP connector would use the correct one.

Also fixed some outdated documentation about a couple of services
that were removed long ago.

Tested with unit tests and by running spark-shell with SSL configs.

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -1781,11 +1781,25 @@ Apart from these, the following properties are also available, and may be useful
...
+            same service. A value of "0" will make the service bind to an ephemeral port.
+        </td>
+    </tr>
+    <tr>
         <td><code>spark.ssl.enabledAlgorithms</code></td>
         <td>Empty</td>
         <td>
             A comma separated list of ciphers. The specified ciphers must be supported by JVM.
-            The reference list of protocols one can find on
+            The reference list of protocols one can found on
```
Comment (ajbozarth) [inline docs/configuration.md]: pretty sure it was grammatically correct before

**Review by ajbozarth — COMMENTED** [review]
Code looks good, just one nit in the doc, but someone with more ssl knowledge should also look

File: docs/configuration.md (hunk 1 above)
Comment (vanzin) [inline docs/configuration.md]: oops.

File: core/src/main/scala/org/apache/spark/SSLOptions.scala [hunk 2]
```diff
@@ -164,6 +167,11 @@ private[spark] object SSLOptions extends Logging {
   def parse(conf: SparkConf, ns: String, defaults: Option[SSLOptions] = None): SSLOptions = {
     val enabled = conf.getBoolean(s"$ns.enabled", defaultValue = defaults.exists(_.enabled))
 
+    val port = conf.getOption(s"$ns.port").map(_.toInt)
+    port.foreach { p =>
+      require(p >= 0, "Port number must be a positive value.")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/SSLOptions.scala]: nit. `positive` -> `non-negative`?

**Comment by vanzin** [comment]
@sarutak want to look at this since you looked at the somewhat mildly related PR?

**Comment by sarutak** [comment]
I'll take a look at this during this week.

File: docs/security.md [hunk 3]
```diff
@@ -49,10 +49,6 @@ component-specific configuration namespaces used to override the default setting
     <th>Component</th>
   </tr>
   <tr>
-    <td><code>spark.ssl.fs</code></td>
```
Comment (sarutak) [inline docs/security.md]: The namespace `fs` seems to be still referred (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/SecurityManager.scala#L259 ).
We shouldn't remove this description should we?

Also it's chance to modify the description about the namespace in `SecurityManager.scala`.

```
 *  All the SSL settings like `spark.ssl.xxx` where `xxx` is a particular configuration property,
 *  denote the global configuration for all the supported protocols. In order to override the global
 *  configuration for the particular protocol, the properties must be overwritten in the
 *  protocol-specific namespace. Use `spark.ssl.yyy.xxx` settings to overwrite the global
 *  configuration for particular protocol denoted by `yyy`. Currently `yyy` can be only`fs` for
 *  broadcast and file server.
```

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 4]
```diff
@@ -394,8 +410,7 @@ private[spark] object JettyUtils extends Logging {
         val httpsURI = createRedirectURI(scheme, baseRequest.getServerName, securePort,
           baseRequest.getRequestURI, baseRequest.getQueryString)
         response.setContentLength(0)
-        response.encodeRedirectURL(httpsURI)
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Nice catch.

**Review by sarutak — COMMENTED** [review]
I tested this change with SparkUI, Standalone and HistoryServer. It almost LGTM. I've left a few comments.

File: docs/configuration.md [hunk 5]
```diff
@@ -1797,6 +1797,20 @@ Apart from these, the following properties are also available, and may be useful
         </td>
     </tr>
     <tr>
+        <td><code>spark.ssl.[namespace].port</code></td>
```
Comment (vanzin) [inline docs/configuration.md]: This was intentional. "spark.ssl.port" doesn't make that much sense if you think about it; you want things like the master UI and history server UI to have different, well known ports, so having this shared config key here doesn't make a lot of sense. For the other configs, such as algorithms and keystore locations, sharing configuration is ok.

File: docs/security.md (hunk 3 above)
Comment (vanzin) [inline docs/security.md]: Hmmm... that code is actually being used, but not to set up the file server, but to configure HTTP clients that download from SSL-enabled servers. Let me see about making that clear in the configuration docs.

**Comment by vanzin** [comment]
Unless I hear back I'll push this tomorrow morning.

**Comment by sarutak** [comment]
LGTM. Merging into `master`. Thanks!

**Pull request closed without merging**